runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -3968,160 +3968,324 @@ inline void generateInlinedCheckCastForDynamicCastClass(TR::Node* node, TR::Code
+            generateLabelInstruction(TR::InstOpCode::JMP4, node, endLabel, cg);
+            }
 
-      // Obtain I-Table
-      generateRegMemInstruction(TR::InstOpCode::LRegMem(), node, itable, generateX86MemoryReference(j9class, offsetof(J9Class, iTable), cg), cg);
-      if (tmp)
-         {
-         generateRegImm64Instruction(TR::InstOpCode::MOV8RegImm64, node, tmp, (uintptr_t)clazz, cg, TR_ClassAddress);
+         og.endOutlinedInstructionSequence();
          }
 
-      // Loop through I-Table
-      generateLabelInstruction(TR::InstOpCode::label, node, iTableLoopLabel, cg);
-      generateRegRegInstruction(TR::InstOpCode::TESTRegReg(), node, itable, itable, cg);
-      generateLabelInstruction(TR::InstOpCode::JE4, node, iTableLookUpFailLabel, cg);
-      auto interfaceMR = generateX86MemoryReference(itable, offsetof(J9ITable, interfaceClass), cg);
-      if (tmp)
-         {
-         generateMemRegInstruction(TR::InstOpCode::CMP8MemReg, node, interfaceMR, tmp, cg);
-         }
-      else
+      // Succeed
+      if (!isCheckCast)
          {
-         generateMemImmSymInstruction(TR::InstOpCode::CMP4MemImm4, node, interfaceMR, (uintptr_t)clazz, node->getChild(1)->getSymbolReference(), cg);
+         generateInstruction(TR::InstOpCode::STC, node, cg);
          }
-      generateRegMemInstruction(TR::InstOpCode::LRegMem(), node, itable, generateX86MemoryReference(itable, offsetof(J9ITable, next), cg), cg);
-      generateLabelInstruction(TR::InstOpCode::JNE4, node, iTableLoopLabel, cg);
+      }
+   else
+      {
+      /**
+       * Do not use a cache.  Generate the interface lookup inline and the fail
+       * path out-of-line.
+       */
+
+      /**
+       * Use the scratch register if it is available.  Otherwise, re-use the
+       * instanceClassReg register to perform itable lookup
+       */
+      TR::Register *itableReg = scratchReg ? scratchReg : instanceClassReg;
+
+      // Obtain I-Table
+      generateRegMemInstruction(
+         TR::InstOpCode::LRegMem(),
+         node,
+         itableReg,
+         generateX86MemoryReference(instanceClassReg, offsetof(J9Class, iTable), cg),
+         cg);
+
+      inlineInterfaceLookup(
+         node,
+         cg,
+         isCheckCast ? iTableLookUpFailLabel : endLabel,
+         castClass,
+         castClassReg,
+         itableReg);
 
-      // Found from I-Table
-      generateMemInstruction(TR::InstOpCode::POPMem, node, generateX86MemoryReference(cache, cg), cg); // j9class
       if (!isCheckCast)
          {
+         // Class found in itable
          generateInstruction(TR::InstOpCode::STC, node, cg);
-         }
-      generateLabelInstruction(TR::InstOpCode::JMP4, node, endLabel, cg);
 
-      // Not found
-      generateVFPRestoreInstruction(vfp, node, cg);
-      generateLabelInstruction(TR::InstOpCode::label, node, iTableLookUpFailLabel, cg);
-      if (isCheckCast)
+         // Fall through to endLabel
+         }
+      else
```

Comment:
I wonder if the case (`not found && !isCheckCast`) is missing here. 

```
      if (!isCheckCast)
         {
         // Class found in itable
         generateInstruction(TR::InstOpCode::STC, node, cg);
         // Fall through to endLabel
         }
      else
         { 
         // isCheckCast && not found 
         // What about (!isCheckCast && not found)?
          ...
         TR_OutlinedInstructionsGenerator og(iTableLookUpFailLabel, node, cg);
         ... 
         }
```

I wonder if the logic might be like this? 
```
      if (!isCheckCast)
         {
         // Class found in itable
         generateInstruction(TR::InstOpCode::STC, node, cg);

         // Fall through to endLabel
         }

      // iTable fail lookup out-of-line
      TR_OutlinedInstructionsGenerator og(iTableLookUpFailLabel, node, cg);
      if (isCheckCast)
         { .... }
      else
          { ... } 

      og.endOutlinedInstructionSequence();
```